Roam: ENG-733 Glue Settings UI components to database operations

sid597 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - End-to-end sync of Roam discourse nodes to the database, including embeddings and concepts.
  - Per-type node retrieval with richer node data (author, type, optional title).
  - Progress reporting during embedding fetches.

- Performance
  - Larger embedding batch size for faster processing.
  - Targeted updates by detecting node types with recent changes.

- Reliability
  - Improved orphaned data cleanup for nodes and schemas.
  - More robust error handling and task coordination during sync.

- Developer Experience
  - Clearer, configurable embedding model usage.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sid597 committed:
sync

sid597 committed:
address coderabbit code

sid597 commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+export const getDiscourseNodeTypeWithSettingsBlockNodes = (
+  node: DiscourseNode,
+  sinceMs: number,
+  extensionAPI: OnloadArgs["extensionAPI"],
```
Will update this to use configTree when this is merged to main and rebase the UI branch off it

mdroidian commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+         [(> ?nodeEditTime ?since)]]
+        ]`;
+
+  const blockNode = window.roamAlphaAPI.data.q(
```
use async https://roamresearch.com/#/app/developer-documentation/page/VJfOx4mgR

sid597 replied:
Oh ok I have been using `data.backend.q`

mdroidian commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+    console.error("No Supabase context found.");
+    return;
+  }
+  const { spaceId, userId } = context;
```
unused

mdroidian commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+  let nodesWithEmbeddings: RoamDiscourseNodeData[];
+  try {
+    nodesWithEmbeddings = await fetchEmbeddingsForNodes(roamNodes);
+  } catch (error: any) {
```
Don't use any if you can avoid it

mdroidian commented on apps/roam/src/utils/fetchEmbeddingsForNodes.ts:
```diff
@@ -19,13 +11,18 @@ type EmbeddingApiResponse = {
 
   for (let i = 0; i < allNodesTexts.length; i += EMBEDDING_BATCH_SIZE) {
     const batch = allNodesTexts.slice(i, i + EMBEDDING_BATCH_SIZE);
+    console.log(
```
Put this behind a dev flag if want to console log it

sid597 replied:
removed

mdroidian commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
-import getDiscourseNodes from "./getDiscourseNodes";
+import getDiscourseNodes, { DiscourseNode } from "./getDiscourseNodes";
```
add `/* eslint-disable @typescript-eslint/naming-convention */` to the top of the file

mdroidian commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -0,0 +1,398 @@
+  SupabaseContext,
+} from "./supabaseContext";
+import { fetchEmbeddingsForNodes } from "./fetchEmbeddingsForNodes";
+import { LocalContentDataInput } from "@repo/database/inputTypes";
```
Are you seeing these errors in your IDE?
<img width="1088" height="258" alt="image" src="https://github.com/user-attachments/assets/d5fb24e2-e651-4064-bf15-41237f97d511" />

maparent replied:
Hrrrm. this used to work, and got broken with changes to tsconfig, package.json etc.
Two problems here:
1. inputTypes is missing from exports in database/package.json, easy to fix. (NOTE: ALL BELOW WAS A MISTAKE, see next message.)
3. Once this is done, the linter complains (rightfully) about roam being in commonJs, and trying to import an ecmaScript module. 

So... what are our options?

1. Use ecmaScript everywhere.  @mdroidian  I tried to just use ecmaScript for roam, and no easy success, i.e. it builds but does not run. Now, that was a minimal attempt, do we have reasons to believe whether or not it could be made to work?
2. use commonJS everywhere, which is a step back.
4. wrap our ecmascript common code in a commonJs wrapper. Found this [option](https://stackoverflow.com/a/75348391).
5. Use UMD everywhere. Rabbit hole for a legacy format imho
6. Not reuse any component from outside in roam. Sigh.
I think this is a blocker that needs to be addressed. Willing to spend more time on this.

maparent replied:
Argh. Forgot to `npm install` in my workspace first, I my workspace was in the `main` state. So all I wrote (after the first 1.) does not apply, i.e. I do not see the second error, though I'm not sure why not. (I also tried rebasing on main, no difference.)
But... I saw those errors before and now, and I'm sure there are conditions that trigger them, so will leave this comment here for now.

sid597 replied:
Yeah and I don't know how to solve it the right way, these I think also lead to lint errors in `convertDgToSupabaseConcepts` function

maparent replied:
```
--- a/packages/database/package.json
+++ b/packages/database/package.json
@@ -7,6 +7,7 @@
   "exports": {
     "./types.gen.ts": "./types.gen.ts",
     "./types.gen": "./types.gen.ts",
+    "./inputTypes": "./inputTypes.ts",
     "./dbDotEnv": "./dbDotEnv.ts"
   },
   "scripts": {
```

mdroidian reviewed (changes requested):
Let's try to clean up as much of the warning/errors as possible in a first round pass.

If you aren't seeing these, let me know
<img width="1415" height="396" alt="image" src="https://github.com/user-attachments/assets/4266d8ab-093c-4550-96dc-2775c632791d" />

<img width="1117" height="313" alt="image" src="https://github.com/user-attachments/assets/f0873cc9-64c6-4b5b-9857-5f53192c242a" />

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+
+
+const EMBEDDING_BATCH_SIZE = 200;
+const API_URL = `https://discoursegraphs.com/api/embeddings/openai/small`;
```
I suggest using `import { nextApiRoot } from "@repo/ui/lib/execContext";` to get the start of this URL.

maparent replied:
(this and other comments will move to fetchEmbeddingsForNodes)

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+      body: JSON.stringify({ input: batch }),
+    });
+
+    if (!response.ok) {
```
At some point there were a lot of transient errors with openAI, and retrying was expected. Do you know if this is still the case?
One case where I know this would still be true is if we ever run into request throttling, in which case we probably want incremental backoff. I know we're far from that much usage, but that will become an issue with more adopters. Punting that should at least be a conscious decision.

sid597 replied:
No I am not aware if its still the case. We can add this as a comment in code for later, wdyt @mdroidian

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+
+    const data: EmbeddingApiResponse = await response.json();
+    if (!data || !Array.isArray(data.data)) {
+      throw new Error(
```
also may be worth retrying.

sid597 replied:
should we not do it on server side?

maparent replied:
That does make sense. I suggest adding a linear task for that.

sid597 replied:
[ENG-761: Retry getting embedding for batches on the api/embeddings/openai/small route](https://linear.app/discourse-graphs/issue/ENG-761/retry-getting-embedding-for-batches-on-the-apiembeddingsopenaismall)

maparent commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
-type RoamDiscourseNodeData = {
+type ISODateString = string;
+
+export type RoamDiscourseNodeData = {
```
~Could it be simpler to use LocalContentDataInput directly? (I tried and no major obstacle. Actually used LocalContentDataInput | null)~
Actually, you're right, use this as a the roam function output, minus the vector. Then convert to `LocalContentDataInput` ~as the output of~ before sending to `fetchEmbeddingsForNodes`.

maparent commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+          type: node.type,
+        };
+      })
+      .filter((n): n is RoamDiscourseNodeData => n !== null),
```
Not sure I understand the syntax here,  how is it different from simply `(n) => n !== null`

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+  }[];
+};
+
+export const fetchEmbeddingsForNodes = async (
```
This looks almost identical to the code in `fetchEmbeddingsForNode`, is there a reason you cannot merge them?

sid597 replied:
my bad

maparent commented on apps/roam/src/utils/fetchEmbeddingsForNodes.ts:
```diff
@@ -19,13 +11,18 @@ type EmbeddingApiResponse = {
-  nodes: DiscourseGraphContent[],
-): Promise<DiscourseGraphContent[]> => {
+  nodes: RoamDiscourseNodeData[],
+): Promise<RoamDiscourseNodeData[]> => {
```
So I suggest using `LocalContentDataInput` as the *input and* output type here

sid597 replied:
done

maparent commented on apps/roam/src/utils/fetchEmbeddingsForNodes.ts:
```diff
@@ -67,7 +64,7 @@ export const fetchEmbeddingsForNodes = async (
   }
   return nodes.map((node, i) => ({
     ...node,
-    model: "openai_text_embedding_3_small_1536",
```
and here wrap those two in the `embedding_inline`. (Annoyingly TS does not flag errors in the map.)

sid597 commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -0,0 +1,398 @@
+        shouldProceed: false,
+      };
+    }
+    const worker = window.roamAlphaAPI.user.uid();
```
This also throws errors but also works

maparent replied:
I think you mean shows linting error? ("Throwing" suggests something else!) But yes, there is a mismatch here.

maparent replied:
(I mean the type we receive is incomplete.)

sid597 replied:
yeah linting error, we don't have `roamAlphaAPI.user.uid` in the definition but roam supports it

<img width="750" height="412" alt="Image" src="https://github.com/user-attachments/assets/d2fc0425-9dab-4a1e-926d-ee4f5cd98465" />

sid597 committed:
address lint errors

sid597 committed:
use async instead of backend

maparent commented on apps/roam/src/utils/fetchEmbeddingsForNodes.ts:
```diff
@@ -19,13 +11,18 @@ type EmbeddingApiResponse = {
+): Promise<RoamDiscourseNodeData[]> => {
   const allEmbeddings: number[][] = [];
-  const allNodesTexts = nodes.map((node) => node.text);
+  const allNodesTexts = nodes.map((node) =>
```
And ideally that transformation should be part of a separate utility function that translates `RoamDiscourseNodeData` to `LocalContentDataInput`, setting the variant at the same time. Having this as a separate type-shifting function would ensure we'd never forget to do it.

sid597 replied:
Done

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,168 @@
+  }
+};
+
+export const upsertNodesToSupabaseAsContentWithEmbeddings = async (
```
Exists here and in syncDgNodesToSupabase. Unify.

maparent commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -0,0 +1,398 @@
+  }
+};
+
+export const convertDgToSupabaseConcepts = async (
```
This converts _and uploads_. I think the name should reflect this. (Or better still, have this function return the converted concepts, and another function that calls this and upload. It'll be easier to test the former.)

maparent commented on apps/roam/src/utils/syncDgNodesToSupabase.ts:
```diff
@@ -0,0 +1,398 @@
+      await endSyncTask(worker, "failed");
+      return;
+    }
+    await upsertNodesToSupabaseAsContentWithEmbeddings(
```
So ideally the nodes should be converted to `LocalContentDataInput` before calling this function and the next one, it will simplify both.

sid597 replied:
we need to know the discourse node schema type for the instance node see line 223 and the `RoamDiscourseNodeData` has that so we cannot pass the converted type

maparent reviewed (changes requested):
I want to see what it looks like with a better type boundary, and then I'll look again at some of the update logic (but that looks good overall.)

sid597 committed:
address review

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

maparent reviewed (approved):
Good. I'll trust you checked the runtime. LGTM.

sid597 committed:
bug fixes

sid597 committed:
bug fixes

sid597 committed:
Merge branch 'eng-733-glue-ui-components-to-data-extraction-and-database' of https://github.com/DiscourseGraphs/discourse-graph into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
unused import

maparent commented on apps/roam/src/utils/upsertNodesAsContentWithEmbeddings.ts:
```diff
@@ -0,0 +1,172 @@
+      ? `${node.node_title} ${node.text}`
+      : node.text;
+    return {
+      author_id: userId,
```
Actually... this is wrong, as it makes the current user the author of all nodes. In theory, you should not have to set this, since you set `author_local_id`; but in practice, the author_local_id may not have been materialized in the database, which may lead to the insert failing. So make sure to collect all author_ids and creator_ids, and to create platformAccounts (prob. with createAccountInSpace; I will create a bulk version of that.)

sid597 replied:
I did not see this comment before I need to learn the github ui better

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

maparent committed:
Add an input type for platform accounts
Add functions to upsert platform accounts (individually or in bulk)
Keep the old create_account_in_space function as a shim for now.
Use the new upsert_account in upsert_documents and upsert_content,
allowing for more complete inline information.

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' of https://github.com/DiscourseGraphs/discourse-graph into eng-764-bulk-account-uploading

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' of https://github.com/DiscourseGraphs/discourse-graph into eng-764-bulk-account-uploading

sid597 committed:
Merge branch 'eng-764-bulk-account-uploading' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
bulk upsert accounts, use database function imports

sid597 committed:
Merge branch 'main' into eng-733-glue-ui-components-to-data-extraction-and-database

sid597 committed:
add comment for future

maparent reviewed (approved):
Ok, this basically lgtm. One small nit before you push: There are a lot of imports that could be `import type {...} ...`, or at least `import {x, type y} ...`, fwiw.

sid597 committed:
use better import method

sid597 committed:
commonjs-to-esm

mdroidian commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+]`;
+
+  //@ts-ignore - backend to be added to roamjs-components
+  const allNodes = (await window.roamAlphaAPI.data.async.q(
```
@sid597 This should probably be `.backend`.   We can discuss at our next meeting.

mdroidian commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+         [(> ?nodeEditTime ?since)]]
+        ]`;
+
+  const blockNode = window.roamAlphaAPI.data.q(
```
@sid597 This should probably be `.backend`.   We can discuss at our next meeting.

mdroidian commented on apps/roam/src/utils/getAllDiscourseNodesSince.ts:
```diff
@@ -1,48 +1,171 @@
+      if (!entity.source_local_id) {
+        return [];
+      }
+      const node = findDiscourseNode(entity.source_local_id, discourseNodes);
```
@sid597 This will likely need to be a custom function.  `findDiscourseNode` will likely be a performance issue on large graphs.  We can discuss at our next meeting.
